burn-core/src/optim/rmsprop.rs

```diff
@@ -0,0 +1,528 @@
+/// The optimizer can be configured with [RMSPropConfig](RMSPropConfig).
+pub struct RMSProp<B: Backend> {
+    alpha: f32,
+    // epsilon: f32,
+    centered: bool,
+    // momentum: Option<Momentum<B>>,
+    momentum: RMSPropMomentum,
+    weight_decay: Option<WeightDecay<B>>,
+}
+
+/// State of [RMSProp](RMSProp)
+#[derive(Record, Clone, new)]
+pub struct RMSPropState<B: Backend, const D: usize> {
+    weight_decay: Option<WeightDecayState<B, D>>,
+    square_avg: Option<SquareAvgState<B, D>>,
+    centered: Option<CenteredState<B, D>>,
+    momentum: Option<RMSPropMomentumState<B, D>>,
+}
+
+/// [SquareAvgState](SquareAvgState) is to store and pass optimizer step params.
+#[derive(Record, Clone, new)]
+pub struct SquareAvgState<B: Backend, const D: usize> {
+    square_avg: Tensor<B, D>,
+}
+
+impl<B: Backend, const D: usize> SquareAvgState<B, D> {
+    /// transform [SquareAvgState] to the next step
+    pub fn transform(alpha: f32, grad: Tensor<B, D>, state: Option<Self>) -> (Tensor<B, D>, Self) {
+        match state {
+            Some(state) => {
+                let square_avg = state
+                    .square_avg
+                    .clone()
+                    .mul_scalar(alpha)
+                    .add(grad.clone().powf(2.).mul_scalar(1. - alpha));
+                (grad, Self { square_avg })
+            }
+            _ => {
+                let square_avg = grad.clone().powf(2.).mul_scalar(1. - alpha);
+                (grad, Self { square_avg })
+            }
+        }
+    }
+
+    fn to_device(mut self, device: &B::Device) -> Self {
+        self.square_avg = self.square_avg.to_device(device);
+        self
+    }
+}
+
+/// [CenteredState](CenteredState) is to store and pass optimizer step params.
+#[derive(Record, Clone, new)]
+pub struct CenteredState<B: Backend, const D: usize> {
+    grad_avg: Tensor<B, D>,
+    avg: Tensor<B, D>,
+}
+
+impl<B: Backend, const D: usize> CenteredState<B, D> {
+    /// transform [CenteredState] to the next step
+    pub fn transform(
+        alpha: f32,
+        centered: bool,
+        grad: Tensor<B, D>,
+        centered_state: Option<Self>,
+        square_avg_state: Option<SquareAvgState<B, D>>,
+    ) -> (Tensor<B, D>, Self, SquareAvgState<B, D>) {
+        if centered {
+            let grad_avg = match centered_state {
+                Some(state) => state
+                    .grad_avg
+                    .clone()
+                    .mul_scalar(alpha)
+                    .add(grad.clone().mul_scalar(1. - alpha)),
+                _ => grad.clone().mul_scalar(1. - alpha),
+            };
+            let (avg, square_avg) = match square_avg_state {
+                Some(state) => (
+                    state.square_avg.clone().sub(grad_avg.clone().powf(2.)),
+                    state.square_avg,
+                ),
+                _ => {
+                    panic!("uninitialized square_avg state")
+                }
```

Comment:
We should avoid this by making `square_avg_state` not optinal.

Reply:
Done.